### PR TITLE
Add user home mount for ephemeral storage when init-persistent-home i…

### PR DESCRIPTION
**Pull request by dkwon17**
…s explicitly configured

### What does this PR do?
If an init-persistent-home init container is explicitly defined in the DWOC's `config.workspaces.initContainers` field and if persistUserHome is enabled, a /user/home empty dir volume mount will be added to the workspace pod.

This is to allow consistent behaviour between ephemeral and non-ephemeral workspaces when `persistUserHome` is enabled and a custom `init-persistent-home` init container is defined.

### What issues does this PR fix or reference?
https://redhat.atlassian.net/browse/CRW-11010

### Is it tested? How?

Install DWO with the following catalog source:
```
apiVersion: operators.coreos.com/v1alpha1
  kind: CatalogSource
  metadata:
    name: dwo-custom
    namespace: openshift-marketplace
  spec:
    sourceType: grpc
    image: quay.io/dkwon17/devworkspace-operator-index:ephemeral-mount
    displayName: DWO Custom
    updateStrategy:   
      registryPoll:
        interval: 5m
```

For both cases, ensure that the global DWOC has `persistUserHome` enabled:
```
apiVersion: controller.devfile.io/v1alpha1
  kind: DevWorkspaceOperatorConfig
  metadata:
    name: devworkspace-operator-config
    namespace: openshift-operators
  config:   
    workspace:
      persistUserHome:
        enabled: true
```

<details><summary>Test case 1</summary>

1. Create an Empty Workspace with ephemeral storage:
```
apiVersion: workspace.devfile.io/v1alpha2
  kind: DevWorkspace
  metadata:
    name: my-workspace
  spec:
    started: true
    template:
      attributes:
        controller.devfile.io/storage-type: ephemeral
      components:
        - name: tooling-container
          container:
            image: quay.io/devspaces/udi-rhel9:latest
```
2. Verify pod has no `init-persistent-home` init container and no PVC created

</details>

<details><summary>Test case 2</summary>

1. Update global DWOC such that `init-persistent-home` explicitly configured:
```
apiVersion: controller.devfile.io/v1alpha1
  kind: DevWorkspaceOperatorConfig
  metadata:
    name: devworkspace-operator-config
    namespace: openshift-operators
  config:
    workspace:
      initContainers:
        - name: init-persistent-home
          image: registry.access.redhat.com/ubi9/ubi-minimal:latest
          command: ["/bin/sh", "-c", "echo custom-init-ran"]
```

2. Create an Empty Workspace with ephemeral storage:
```
apiVersion: workspace.devfile.io/v1alpha2
  kind: DevWorkspace
  metadata:
    name: my-workspace
  spec:
    started: true
    template:
      attributes:
        controller.devfile.io/storage-type: ephemeral
      components:
        - name: tooling-container
          container:
            image: quay.io/devspaces/udi-rhel9:latest
```

3. Wait until workspace starts. Verify `init-persistent-home` ran, `persistent-home` volume is `emptyDir` (not PVC), no PVC created.

For example:

image

image

</details>

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

## Release Notes

* **New Features**
  * Refined persistent home eligibility rules, including improved handling for ephemeral workspaces and stricter checks for existing `/home/user/` mounts and initialization configuration.

* **Tests**
  * Added/updated test scenarios covering ephemeral storage with and without `init-persistent-home`, and improved assertions for expected failure messaging and resulting workspace components.

### CODE REVIEW

**Comment by tolusha** [comment]
Hi! I'm **che-ai-assistant** — I help with your pull requests.

**Available commands**:
- `/che-ai-assistant generate-che-doc` — Generate a documentation PR based on this PR's changes
- `/che-ai-assistant ok-pr-review` — Run a comprehensive PR review (summary, code review, deep review, impact analysis)
- `/che-ai-assistant help` — Show this help message

**Comment by rohanKanojia** [comment]
/che-ai-assistant ok-pr-review

Review is complete. Please check the review comments below.

**Review by tolusha — APPROVED** [review]
## Code Review Summary

### 🎯 Overall Assessment: **APPROVE** ✅

This PR successfully implements support for persistent home directories in ephemeral storage workspaces when a custom `init-persistent-home` init container is explicitly configured. The changes are well-structured, properly tested, and solve a real consistency problem between ephemeral and non-ephemeral workspaces.

---

### ✅ Strengths

1. **Clean Architecture** - Removed coupling with `storage` package, making the home logic more self-contained and maintainable
2. **Proper Test Coverage** - Added 2 comprehensive test cases covering both new behavior and backward compatibility
3. **Clear Logic** - The special handling for ephemeral storage is explicit and easy to understand
4. **Backward Compatible** - Existing workspaces without custom init containers maintain current behavior
5. **Good Error Fix** - Corrected error message from "command" to "event" for accuracy

---

### 🔍 Key Observations

#### 1. Behavioral Change for Non-Ephemeral Storage (Medium Priority)

The removal of `storage.WorkspaceNeedsStorage()` check simplifies the logic but changes behavior:

**Before:**
```go
return storage.WorkspaceNeedsStorage(&workspace.Spec.Template)
```

**After:**
```go
return true  // for all non-ephemeral storage
```

**Impact:** A workspace with non-ephemeral storage type but only ephemeral volumes will now get persistent home (previously wouldn't).

**Question:** Is this intentional? If so, consider adding a clarifying comment:

```go
// For ephemeral storage, only add persistent home if init-persistent-home is configured
storageType := workspace.Spec.Template.Attributes.GetString(constants.DevWorkspaceStorageTypeAttribute, nil)
if storageType == constants.EphemeralStorageClassType {
    return hasInitPersistentHomeInConfig(workspace)
}

// For non-ephemeral storage types, always create persistent home if enabled.
// This includes workspaces with all ephemeral volumes but non-ephemeral storage type.
return true
```

#### 2. Container Component Validation (Low Priority)

Good addition of the `hasContainerComponents` check - prevents edge cases where workspaces without containers would incorrectly trigger home creation.

---

### 💡 Suggestions for Improvement

1. **Additional Test Case** (Optional)
   - Test workspace with non-ephemeral storage type but all volumes marked ephemeral
   - Validates the behavioral change mentioned above

2. **Documentation Update** (Recommended)
   - Update operator docs to explain the new ephemeral + custom init behavior
   - Add example configuration showing this capability

---

### 📋 Pre-Merge Checklist

- [x] Code quality - Clean, maintainable changes
- [x] Unit tests - New tests added and passing
- [x] Logic correctness - Functions as designed
- [ ] E2E tests - Waiting for `/test v8-devworkspace-operator-e2e, v8-che-happy-path`
- [ ] Manual testing - Per test plan in PR description
- [ ] Documentation - Consider updating user-facing docs

---

### 🎬 Recommendation

**Approve and merge** after E2E tests pass. The code changes are solid, well-tested, and implement the requested functionality correctly.

### Technical Details Reviewed

- ✅ Logic refactoring in `NeedsPersistentHomeDirectory()`
- ✅ New helper function `hasInitPersistentHomeInConfig()`
- ✅ Test coverage for new behavior
- ✅ Backward compatibility maintained
- ✅ Error message correction
- ✅ Dependency reduction (removed storage import)

---

**Review conducted:** 2026-06-17
**Files analyzed:** 4 files (126 changed lines + context)
**Test coverage:** 2 new test cases added
**Risk assessment:** Low-Medium (E2E validation recommended)

**Comment by rohanKanojia** [comment]
I tested as per abovementioned comment and can confirm it's working as expected:

When no init-persistent-home init container, ephemeral workspace starts without any init container

```
workspaceb387c01a8fb64a18-7ccd758878-h5r8t   0/1     ContainerCreating   0          2s
workspaceb387c01a8fb64a18-7ccd758878-h5r8t   1/1     Running             0          4s
```

When init-persistent-home init container is in DWOC, workspace has extra init `init-persistent-home` container 

```
workspace8828788e71074dbc-8df88b6b6-hmwpj   0/1     Init:0/1      0          0s
workspace8828788e71074dbc-8df88b6b6-hmwpj   0/1     PodInitializing   0          19s
workspace8828788e71074dbc-8df88b6b6-hmwpj   1/1     Running           0          22s
```

File: pkg/library/home/testdata/persistent-home/noop-if-init-prestartevent-already-defined.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ input:
         - init-persistent-home
 
 output:
-  error: "failed to add init container for home persistence setup: command with id init-persistent-home already exists in the devworkspace"
+  error: "failed to add init container for home persistence setup: event with id init-persistent-home already exists in the devworkspace"
```
Comment (rohanKanojia) [inline pkg/library/home/testdata/persistent-home/noop-if-init-prestartevent-already-defined.yaml:19]: Umm, why is this change required?

**Review by rohanKanojia — APPROVED** [review]

File: pkg/library/home/testdata/persistent-home/noop-if-init-prestartevent-already-defined.yaml (hunk 1 above)
Comment (dkwon17) [reply pkg/library/home/testdata/persistent-home/noop-if-init-prestartevent-already-defined.yaml:19]: Good catch, this change is not necessary

Comment (dkwon17) [reply pkg/library/home/testdata/persistent-home/noop-if-init-prestartevent-already-defined.yaml:19]: Actually, I noticed a problem. in this test case, the noop happens because there are no container components (ex. [1]). My latest commit adds a container component to to the test case (just like test cases [2], [3]) so that the current test case can verify that the no op happens due to the `init-persistent-home` event already existing. 

[1] https://github.com/devfile/devworkspace-operator/blob/main/pkg/library/home/testdata/persistent-home/noop-if-no-components.yaml
[2] https://github.com/devfile/devworkspace-operator/blob/main/pkg/library/home/testdata/persistent-home/noop-if-init-command-already-defined.yaml
[3] https://github.com/devfile/devworkspace-operator/blob/main/pkg/library/home/testdata/persistent-home/noop-if-init-component-already-defined.yaml

Comment (rohanKanojia) [reply pkg/library/home/testdata/persistent-home/noop-if-init-prestartevent-already-defined.yaml:19]: thanks for detailed explanation! It can be resolved.
